python/dlisio/__init__.py

```diff
@@ -749,70 +749,73 @@ def load(path):
-        hint = rewind(stream.absolute_tell, tapemarks)
-
-        records = core.extract(stream, explicits)
-        fdata_index = defaultdict(list)
-        for key, val in core.findfdata(stream, implicits):
-            fdata_index[key].append(val)
-
-        lf = dlis(stream, explicits, records, fdata_index, sul)
-        lfs.append(lf)
-
         try:
-            stream = core.open(path)
-            offset = core.findvrl(stream, hint)
-        except RuntimeError:
-            if stream.eof():
-                stream.close()
-                break
-            raise
-
-    return Batch(lfs)
+            offset = core.findsul(stream)
+            sul = stream.get(bytearray(sulsize), offset, sulsize)
+            offset += sulsize
+        except:
+            offset = 0
+            sul = None
+
+        tapemarks = core.hastapemark(stream)
+        offset = core.findvrl(stream, offset)
+
+        def rewind(offset, tif):
+            """Rewind offset to make sure not to miss VRL when calling findvrl"""
+            offset -= 4
+            if tif: offset -= 12
+            return offset
+
+        # Layered File Protocol does not currently offer support for re-opening
+        # files at the current position, nor is it able to precisly report the
+        # underlying tell. Therefore, dlisio has to manually search for the
+        # VRL to determine the right offset in which to open the new filehandle
+        # at.
+        #
+        # Logical files are partitioned by core.findoffsets and it's required
+        # [1] that new logical files always start on a new Visible Record.
+        # Hence, dlisio takes the (approximate) tell at the end of each Logical
+        # File and searches for the VRL to get the exact tell.
+        #
+        # [1] rp66v1, 2.3.6 Record Structure Requirements:
+        #     > ... Visible Records cannot intersect more than one Logical File.
+        lfs = []
+        while True:
+            if tapemarks: offset -= tifsize
+            stream.seek(offset)
+            if tapemarks: stream = core.open_tif(stream)
+            stream = core.open_rp66(stream)
+
+            explicits, implicits = core.findoffsets(stream)
+            hint = rewind(stream.absolute_tell, tapemarks)
+
+            records = core.extract(stream, explicits)
+            fdata_index = defaultdict(list)
+            for key, val in core.findfdata(stream, implicits):
+                fdata_index[key].append(val)
+
+            lf = dlis(stream, explicits, records, fdata_index, sul)
+            lfs.append(lf)
+
+            try:
+                stream = core.open(path)
+                offset = core.findvrl(stream, hint)
+            except RuntimeError:
+                if stream.eof(): break
+                raise
+
+        return Batch(lfs)
+
+    finally:
+        # last object assigned to stream variable is always broken or of no
+        # interest to us, so we can safely close it
```

Comment:
This is insufficient whenever something fails. Assume load has successfully gathered up two logical files in `lfs` then the third one fails on e.g. `findoffsets`, now only the last/current one is released, not the two in `lfs`.

This would not be an issue if we actually _returned_  the successfully indexed logical files, so maybe we address that?

Reply:
oh, you are so right. totally missed that :woman_facepalming: